pyomo/gdp/transformed_disjunct.py

```diff
@@ -0,0 +1,55 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright (c) 2008-2022
+#  National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+from pyomo.common.autoslots import AutoSlots
+from pyomo.core.base.block import _BlockData, Block
+from pyomo.core.base.global_set import UnindexedComponent_index
+
+class _TransformedDisjunctData(_BlockData):
+    __slots__ = ('_src_disjunct',)
+    __autoslot_mappers__ = {'_src_disjunct': AutoSlots.weakref_mapper}
+
+    @property
+    def src_disjunct(self):
+        return self._src_disjunct
```

Comment:
If we are changing the API, should this property resolve the weakref?  That is,
```python
    @property
    def src_disjunct(self):
        return None if self._src_disjunct is None else self._src_disjunct()
```

Reply:
It could, but it would be inconsistent with the mirror `transformation_block` property on `Disjunct`, which is why I chose this. I don't have strong feelings though.

Reply:
I am coming around to the opinion that the use of weakrefs in Pyomo is an implementational detail that should not "bleed" into the user experience.  So maybe I would argue for hiding the weakref behind a property in both cases?  Doesn't necessarily have to be this PR, though.

Reply:
OK, I think I agree with you. I'm going to slip it into this PR...